### ISSUE TITLE
equality on different units

### ISSUE THREAD

**Issue by daniv** [dropwizard/dropwizard#2153]
`Assertion.assertThat( reason, CoreMatchers.equalTo( Size.terabytes( 0 ) ), Size.parse( "0M" ) )
java.lang.AssertionError: parse( '0M' ) matches :<0 terabytes>
Expected: <0 terabytes>
     but: was <0 megabytes>`

`Assertion.assertThat( reason, CoreMatchers.equalTo( Size.bytes( 0 ) ), Size.terabytes( 0 ) );
java.lang.AssertionError: Size.bytes( 0 ) matches :<0 terabytes>
Expected: <0 bytes>
     but: was <0 terabytes>
`

looks like 0 ( zero ) no equals to 0 zero.........

**Comment by lutzhorn** [comment]
Yes, because `Size.equals` is

````java

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if ((obj == null) || (getClass() != obj.getClass())) {
            return false;
        }
        final Size size = (Size) obj;
        return (count == size.count) && (unit == size.unit);
    }
````

Both the count and the unit must be equal.

This does not look like a bug to me.

**Comment by nickbabcock** [comment]
500 bytes can't be represented as a whole megabyte, so one would have to round to 0 or 1.

**Comment by isaki-x** [comment]
There may be a nice way to fix this; I do something similar for working with some of my own date and time constructs (for example, 60 seconds and 1 minute are equal in spite of using different units of measurement). I'll take a look at this and see if the logic I utilized is applicable here.

**Comment by isaki-x** [comment]
Actually, this class is actually bugged. It violates the Comparable contract. Two objects may show compareTo == 0 while equals returns false.

**Comment by isaki-x** [comment]
I have added a unit test to my local code base that fails (which should succeed).

```Java
    @Test
    public void verifyComparableContract() {
        final Size kb = Size.kilobytes(1024L);
        final Size bytes = Size.bytes(kb.toBytes());
        
        assertThat(bytes.compareTo(kb)).isEqualTo(0);
        assertThat(kb.compareTo(bytes)).isEqualTo(0);
        
        // If comparator == 0, then the following must be true
        assertThat(bytes.equals(kb)).isTrue();
        assertThat(kb.equals(bytes)).isTrue();
    }
```

I will submit a PR with a fix.

**Comment by joschi** [comment]
Closed via #2155.
